watertap/unit_models/coag_floc_model.py

```diff
@@ -714,6 +1043,97 @@ def calculate_scaling_factors(self):
+
+            for ind, c in self.eq_rapid_mixing_power.items():
+                iscale.constraint_scaling_transform(c, sf)
+
+        # set scaling for floc_basin_vol
+        if iscale.get_scaling_factor(self.floc_basin_vol[t]) is None:
+            sf1 = 0
+            sf2 = 0
+            for t in self.control_volume.properties_out:
+                sf1 += iscale.get_scaling_factor(
+                    self.control_volume.properties_out[t].flow_vol_phase["Liq"]
+                )
+            sf2 = iscale.get_scaling_factor(self.floc_retention_time)
+            sf1 = sf1 / len(self.control_volume.properties_in)
+            sf = sf1 * sf2
+            iscale.set_scaling_factor(self.floc_basin_vol, sf)
+
+            for ind, c in self.eq_floc_basin_vol.items():
+                iscale.constraint_scaling_transform(c, sf)
+
+        # set scaling for floc_wheel_speed
+        if iscale.get_scaling_factor(self.floc_wheel_speed[t]) is None:
+            sf1 = iscale.get_scaling_factor(self.paddle_rotational_speed)
+            sf2 = iscale.get_scaling_factor(self.single_paddle_length)
+            sf = sf1 * sf2 * Constants.pi / 10
+            iscale.set_scaling_factor(self.floc_wheel_speed, sf)
+
+            for ind, c in self.eq_floc_wheel_speed.items():
+                iscale.constraint_scaling_transform(c, sf)
+
+        # set scaling for flocculation_power
+        if iscale.get_scaling_factor(self.flocculation_power[t]) is None:
+            sf1 = iscale.get_scaling_factor(self.floc_wheel_speed)
+            sf2 = iscale.get_scaling_factor(self.vel_fraction)
+            sf3 = 0
+            for t in self.control_volume.properties_out:
+                sf3 += iscale.get_scaling_factor(
+                    self.control_volume.properties_out[t].dens_mass_phase["Liq"]
+                )
+            sf3 = sf3 / len(self.control_volume.properties_in)
+            sf4 = iscale.get_scaling_factor(self.single_paddle_length)
+            sf5 = iscale.get_scaling_factor(self.single_paddle_width)
+            sf6 = iscale.get_scaling_factor(self.num_paddle_wheels)
+            sf7 = iscale.get_scaling_factor(self.num_paddles_per_wheel)
+            sf8 = iscale.get_scaling_factor(self.paddle_drag_coef)
+            sf = 0.5 * sf8 * (sf4 * sf5 * sf6 * sf7) * sf3 * sf2**3 * sf1**3 * 500
+            iscale.set_scaling_factor(self.flocculation_power, sf)
+
+            for ind, c in self.eq_flocculation_power.items():
+                iscale.constraint_scaling_transform(c, sf)
+
+        # set scaling for total_power
+        if iscale.get_scaling_factor(self.total_power[t]) is None:
+            sf1 = iscale.get_scaling_factor(self.flocculation_power)
+            sf2 = iscale.get_scaling_factor(self.rapid_mixing_power)
+            sf = (sf1 + sf2) / 2
+            iscale.set_scaling_factor(self.total_power, sf)
+
+            for ind, c in self.eq_total_power.items():
+                iscale.constraint_scaling_transform(c, sf)
+
```

Comment:
It seems to be a sophisticated method of assigning scaling factors.  Without checking eq details, I would double if there is a chance of sf turning out 0, which may cancel a var or constraint?

Reply:
Since scaling factors are always positive numbers (and these are all built off of other scaling factors, which are in turn built on scaling factors the user provides) it should only every give a `0` for a scaling factor if the user gives a `0` for a scaling factor, which the user should never do.